lib/std/heap.zig

```diff
@@ -591,6 +591,33 @@ pub fn StackFallbackAllocator(comptime size: usize) type {
     };
 }
 
+var page_size: ?usize = null;
+
+pub fn pageSize() usize {
+    if (builtin.cpu.arch.isWasm()) {
+        return 64 * 1024;
+    }
+
+    if (page_size) |pg_size| return pg_size;
+
+    page_size = switch (builtin.os.tag) {
+        .linux => if (builtin.link_libc) @intCast(std.c.sysconf(std.os.linux.SC.PAGESIZE)) else std.os.linux.getauxval(std.elf.AT_PAGESZ),
+        .macos => std.os.darwin.machTaskForSelf().getPageSize() catch 0,
+        .windows => blk: {
+            var info: std.os.windows.SYSTEM_INFO = undefined;
+            std.os.windows.kernel32.GetSystemInfo(&info);
+            break :blk info.dwPageSize;
+        },
+        else => if (builtin.link_libc and @hasDecl(std.c, "getpagesize")) @intCast(std.c.getpagesize()) else switch (builtin.cpu.arch) {
+            .sparc64 => 8 * 1024,
+            else => 4 * 1024,
+        },
+    };
+
+    assert(page_size.? > 0 and page_size.? % 2 == 0);
+    return page_size.?;
+}
+
```

Comment:
May I suggest replacing this code with the following:
```suggestion
// Marked inline to preserve comptime-ness.
pub inline fn pageSize() usize {
    comptime {
        // "Windows CE for ARM920 took advantage of [1KB subpages] and used 1KB "pages". All other
        // flavors of Windows use the native 4KB pages."
        //
        // -- <https://devblogs.microsoft.com/oldnewthing/20210510-00/?p=105200>
        if (builtin.os.tag == .windows and (!builtin.cpu.arch.isARM() or builtin.os.version_range.windows.min.isAtLeast(.xp))) {
            switch (builtin.cpu.arch) {
                .x86, .x86_64 => 4 << 10,
                // missing: SuperH (4 << 10)
                .mips, .mipsel, .mips64, .mips64el => 4 << 10,
                .powerpc, .powerpcle, .powerpc64, .powerpc64le => 4 << 10,
                // missing: DEC Alpha (8 << 10)
                // missing: Itanium (8 << 10)
                .thumb, .thumbeb, .arm, .armeb, .aarch64, .aarch64_be, .aarch64_32 => 4 << 10,
                else => {},
            }
        }

        switch (builtin.cpu.arch) {
            // "[...] each page is sized 64KiB."
            //
            // -- <https://developer.mozilla.org/en-US/docs/webassembly/reference/memory/size>
            .wasm32, .wasm64 => return 64 << 10,
            .x86_64 => switch (builtin.os.tag) {
                // "If you run a 64-bit system [or any other architecture; sic], the page size is 4K
                // and cannot be changed."
                //
                // -- <https://unix.stackexchange.com/a/80736>
                .linux => return 4 << 10,
                else => {},
            },
            else => {},
        }
    }

    return pageSizeRuntime();
}

fn pageSizeRuntime() usize {
    const cache = struct {
        var page_size: std.atomic.Value(usize) = .{.raw =  0};
    };

    const cached_page_size = cache.page_size.load(.Unordered);

    if (cached_page_size != 0) return cached_page_size;

    var page_size: usize = switch (builtin.os.tag) {
        .linux => if (builtin.link_libc) @intCast(std.c.sysconf(std.os.linux.SC.PAGESIZE)) else std.os.linux.getauxval(std.elf.AT_PAGESZ),
        .macos => std.os.darwin.machTaskForSelf().getPageSize() catch 0,
        .windows => blk: {
            var info: std.os.windows.SYSTEM_INFO = undefined;
            std.os.windows.kernel32.GetSystemInfo(&info);
            break :blk info.dwPageSize;
        },
        else => if (builtin.link_libc and @hasDecl(std.c, "getpagesize")) @intCast(std.c.getpagesize()) else 0,
    };

    if (page_size == 0) {
        // `getauxval` failed due to missing startup code, or `getPageSize` failed, or there is no
        // known way of getting the page size. The user may specify a default, otherwise, the
        // program panics.
        if (@hasDecl(root, "std_assumed_page_size")) {
            page_size = root.std_assumed_page_size;
        } else {
            @panic("unknown page size");
        }
    }

    // Either the OS or the user is mistaken.
    assert(page_size % 2 == 0);

    cache.page_size.store(page_size, .Unordered);
    return page_size;
}
```

It has several advantages:
- Atomics are used to prevent an illegal race.
- More page sizes are known at comptime, lessening the performance impact.
- The cached value is no longer accessible throughout `heap.zig`, preventing misuse.
- The failure mode can be specified by the user. (Should this be in `std.Options`?)

I would love to hear everyone's thoughts on getting the page size from `/proc/self/auxv ` on Linux in case `getauxval` fails, such as in shared libraries.

Reply:
> It's probably better to keep it 1 function

Why?

> also I don't think inline is good for this

As far as I'm aware, inline preserves the comptime-ness of arguments and return values. You can see this in action for arguments in `std.mem.writeInt`, for example.

> Why is page size for Windows compile time?

It's comptime-known because there doesn't seem to be a way to change it, it lessens the performance impact, I used the article I linked as a source. Windows <XP on ARM is excluded because Windows CE on ARM seemed to use 1K pages, and we don't have a way to check for Windows CE specifically, so it is checked at runtime.

> Do we even support Zig mips-windows?

Don't know, doesn't matter for this PR.

> Why is root checked for std_assumed_page_size instead of an std option?

I wasn't sure whether `std.Options` was the correct place to put it. Feel free to put it there.

> Checking /proc/self/auxv does sound like a good idea, I did bring this up about getenv for checking the env vars but it was kinda shot down because of a compatibility thing.

That was for `envp`, right? I saw that issue, but it seemed more like a "there are more important things right now" thing, not like a "this is a bad approach" thing.

Reply:
> Why?

Simplicity

> It's comptime-known because there doesn't seem to be a way to change it, it lessens the performance impact, I used the article I linked as a source. Windows <XP on ARM is excluded because Windows CE on ARM seemed to use 1K pages, and we don't have a way to check for Windows CE specifically, so it is checked at runtime.

Makes sense but for simplicity sake, it is probably better to stick with system calls/API's to request the page size. And this doesn't just go for Windows or a specific target but all targets. Despite Intel MacBooks running macOS use x86 processors, they don't use the standard page size. On the other hand, Windows and Linux appear to use the standard. This would mean that we'd have to add a target check which means a bit more complexity. It's better in the long run imo to KISS (keep it simple stupid).

> I wasn't sure whether std.Options was the correct place to put it. Feel free to put it there.

Yeah, I did that.

> That was for envp, right? I saw that issue, but it seemed more like a "there are more important things right now" thing, not like a "this is a bad approach" thing.
 
Yeah, thought if checking proc for this where to be added then it wouldn't be a part of heap.

Reply:
Sorry, but I don't think making it one or two functions changes the simplicity of it. Making the runtime part a separate function is required to propagate the comptime-ness of the result to the caller and simultaneously let the optimizer decide whether to inline the runtime part or not.

> Makes sense but for simplicity sake, it is probably better to stick with system calls/API's to request the page size.

You yourself correctly specified 64K as the page size for WASM at comptime. This is just an extension of that. I don't see how the rest of that paragraph matters for this particular implementation, this is just an optimization that is necessary because, as you know, the page size is requested very frequently. Caching the result of the syscall in an atomic variable is already a good first step, but if the page size can be determined at comptime, as is the case with WASM, Linux on x86_64, and modern versions of Windows, it should be. 

Regardless of how you feel about the complexity of the proposed solution, this is an important optimization. 

I suggest a compromise: If possible, determine the page size at comptime, but include an assertion that it equals the runtime-determined page size in safe modes. That way, bugs cannot be introduced by erroneous page size detections, but unsafe modes benefit from the page size being comptime-known.

Reply:
Well I do a fixed size for WASM because there's no API or system call for retrieving the page size.

For determining a valid page size, target code will have to be updated to find the valid page sizes. It feels a bit out of scope for this PR as I was intending this just to fix the page size issues with the Pi 5 and Asahi Linux.

Reply:
Ok, I implemented the inline stuff and I've somewhat added the comptime optimizations.

Reply:
Thanks! I have just two requests:
- please put the runtime bit of the function into a separate non-inline function and call it from the function with the comptime bit so that the optimizer can decide whether to inline the runtime part or not
- please add sources for where these page sizes come from (just use the links I sent and source the aarch64 macos one) so others can verify them.

Reply:
I added the sources, does adding the function to outside really change the optimization? `cache` should just be acting like a namespace and so when the compiler sees the call it should just optimize to that directly.

Reply:
>  added the sources, does adding the function to outside really change the optimization? 

Sorry, I didn't catch that you added the function namespaced to `cache`. You are right, this shouldn't change anything.